### PR TITLE
Feature/mission download

### PR DESCRIPTION
**Pull request by ariamirzai**
To test, add line "ui->planUI->setCurrentVehicleConnection(vehicleConnection);" beneath line 62 in mainwindow.cpp in ControlTower.

Also, make sure the WayWise-folders in both ControlTower and RCCar point to this branch.

### CODE REVIEW

**Review by marvind — CHANGES_REQUESTED** [review]
Looks good so far! But I am unable to build, because the changes in the abstract WaypointFollower interface are not implemented in all WaypointFollowers implementing it:

`/usr/include/x86_64-linux-gnu/qt5/QtCore/qsharedpointer_impl.h:242:18: error: cannot declare field ‘QtSharedPointer::ExternalRefCountWithContiguousData<GotoWaypointFollower>::data’ to be of abstract type ‘GotoWaypointFollower’
  242 |         NoCVType data;
      |                  ^~~~
In file included from /home/marvind/src/ControlTower/WayWise/communication/vehicleconnections/mavsdkvehicleconnection.h:18,
                 from /home/marvind/src/ControlTower/WayWise/userinterface/flyui.h:13,
                 from /home/marvind/src/ControlTower/WayWise/userinterface/flyui.cpp:5:
/home/marvind/src/ControlTower/WayWise/autopilot/gotowaypointfollower.h:28:7: note:   because the following virtual functions are pure within ‘GotoWaypointFollower’:
   28 | class GotoWaypointFollower : public WaypointFollower`

File: autopilot/multiwaypointfollower.cpp [hunk 1]
```diff
@@ -142,3 +142,8 @@ void MultiWaypointFollower::receiveDistanceOfRouteLeft(double meters)
 {
     emit distanceOfRouteLeft(meters);
 }
+
+QList<PosPoint> MultiWaypointFollower::getCurrentRoute()
+{
+    return mWaypointList;
```
Comment (marvind) [inline autopilot/multiwaypointfollower.cpp:148]: I do not think this is correct, but should be
`return mWaypointFollowerList[mActiveWaypointFollowerID]->getCurrentRoute()`
Please have a chat with @rickaha

File: userinterface/planui.cpp [hunk 2]
```diff
@@ -325,15 +325,13 @@ void PlanUI::on_splitButton_clicked()
 void PlanUI::setCurrentVehicleConnection(const QSharedPointer<VehicleConnection> &currentVehicleConnection)
 {
     mCurrentVehicleConnection = currentVehicleConnection;
+
+    if(!mCurrentVehicleConnection.isNull())
+        ui->downloadCurrentRouteFromVehicleButton->setEnabled(true);
```
Comment (marvind) [inline userinterface/planui.cpp:330]: -> `ui->downloadCurrentRouteFromVehicleButton->setEnabled(!mCurrentVehicleConnection.isNull());`

File: communication/vehicleconnections/mavsdkvehicleconnection.cpp [hunk 3]
```diff
@@ -804,3 +804,68 @@ VehicleConnection::Result MavsdkVehicleConnection::convertResult(mavsdk::Param::
...
+            return "No system connected.";
+        default:
+            ;
+
+    }
+}
+
+QList<PosPoint> MavsdkVehicleConnection::requestCurrentRouteFromVehicle()
+{    
+    auto mission = mavsdk::MissionRaw{mSystem};
```
[the remark this answers is not in the document]
Comment (marvind) [reply communication/vehicleconnections/mavsdkvehicleconnection.cpp:842]: Please double check your changes, 'mission' was meant to be replaced entirely.

[the remark this answers is not in the document]
Comment (ariamirzai) [reply communication/vehicleconnections/mavsdkvehicleconnection.cpp:842]: Replacing 'mission' to later call 'mavsdk::MissionRaw::download_mission()' results in the following error: "call to non-static member function without an object argument".

I've also tried to keep the implementation faithful to the usage example provided by MAVSDK guide (see bottom of page): https://mavsdk.mavlink.io/main/en/cpp/guide/missions.html

Maybe you know a better way to call the function than  'mavsdk::MissionRaw::download_mission()' ?

File: autopilot/multiwaypointfollower.cpp (hunk 1 above)
Comment (ariamirzai) [reply autopilot/multiwaypointfollower.cpp]: Solution corrected after discussion with @rickaha

**Comment by marvind** [comment]
Works like a charm and looks nice. Great job! 🥳
As we are now building with -Wall -Wextra -Wpedantic, I fixed a little warning:
`warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<mavsdk::MissionRaw::MissionItem>::size_type’ {aka ‘long unsigned int’}`

**Review by marvind — APPROVED** [review]

**Pull request merged**